competitors/alex.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <utility>
+#include "../utils/tracking_allocator.h"
+#include "base.h"
+#include "./ALEX/src/core/alex.h"
+#include "./ALEX/src/core/alex_base.h"
+
+template<class KeyType, int size_scale>
+class Alex : public Competitor {
+public:
+  Alex() : map_(TrackingAllocator<std::pair<KeyType, uint64_t>>(
+      total_allocation_size)) {
+  }
+
+  uint64_t Build(const std::vector<KeyValue<KeyType>>& data) {
+    std::vector<std::pair<KeyType, uint64_t>> loading_data;
+    loading_data.reserve(data.size());
+
+    for (auto& itm: data) {
+      uint64_t idx = itm.value;
+      if (size_scale > 1 && idx % size_scale != 0)
+        continue;
+      loading_data.push_back(std::make_pair(itm.key, itm.value));
+    }
+
+    data_size_ = data.size();
+
+    return util::timing([&] {
+      map_.bulk_load(loading_data.data(), loading_data.size());
+    });
+  }
+
+  SearchBound EqualityLookup(const KeyType lookup_key) const {
+    auto it = map_.lower_bound(lookup_key);
+
+    uint64_t guess;
+    if (it == map_.cend()) {
+      guess = data_size_ - 1;
+    } else {
+      guess = it.payload();
+    }
+
+    const uint64_t error = size_scale - 1;
+
+    const uint64_t start = guess < error ? 0 : guess - error;
+    const uint64_t
+        stop = guess + 1 > data_size_ ? data_size_ : guess + 1; // stop is exclusive (that's why +1)
+
+    return (SearchBound){ start, stop };
+  }
+
+  std::string name() const {
+    return "ALEX";
+  }
+
+  std::size_t size() const {
+    return map_.get_allocator().total_allocation_size + sizeof(*this);
```

Comment:
Does ALEX have a `size` method? Otherwise, it's fine to use the tracking allocator approach.

Reply:
The documentation for ALEX has two methods below with size:

```
// Size in bytes of all the model nodes (including pointers) and metadata in data nodes
long long model_size();

// Size in bytes of all the keys, payloads, and bitmaps stored in this index
long long data_size();
```

I didn't know if taking the sum of these would take the total model size, though, and since they also aren't const methods I decided to just use a tracking allocator instead.